### PR TITLE
Comment to help ppl understand.

### PR DESCRIPTION
**Pull request by Endermanbugzjfc**
I found codes in InfoAPI are not easy to read. Reader might confuse with the context even though some places were seem to be self-explanatory.

For example, I spent some time trying understand why had you done this (so I added comment to this line):
https://github.com/SOF3/InfoAPI/blob/d25c1d2d928aa456b163224ae8bc3e4f3fea59bd/src/SOFe/InfoAPI/Ast/InfoSegment.php#L47-L49

### CODE REVIEW

**Comment by Endermanbugzjfc** [comment]
Sorry, forgot to re-fork before starting the branch.

I forked a long time ago. I do not even commit to a fork's default branch nowadays.

File: src/SOFe/InfoAPI/Ast/ChildName.php [hunk 1]
```diff
@@ -53,22 +53,30 @@ static public function parse(string $fqn) : self {
...
+					// Not a subsequence.
 					return false;
 				}
 			}
 			++$match;
 		}
 
+		// Is subsequence that last until the end.
 		return true;
 	}
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/ChildName.php:81]: Actually I am a bit confused about this function now. Are we checking if `$this` is subsequence of `$pattern`, or `$pattern` is subsequence of `$this`? The documentation states the former, but the code appears to be the latter.

File: src/SOFe/InfoAPI/Ast/Expression.php [hunk 2]
```diff
@@ -28,5 +28,8 @@
 final class Expression {
 	public Path $path;
 
+	/**
+	 * (Candidiates of coalescence paths in nest form.)
+	 */
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/Expression.php:33]:
```suggestion
	/** Candidates of coalescence paths in a linked list. */
```

File: src/SOFe/InfoAPI/Ast/InfoSegment.php [hunk 3]
```diff
@@ -31,30 +31,45 @@
  * A segment of brace-enclosed info.
  */
 final class InfoSegment implements Segment {
+	/**
+	 * Paths collapsed nestly.
+	 * (An {@link Expression->alternative} holds the next coalescence path.)
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/InfoSegment.php:36]:
```suggestion
	 * Linked list of the coalescence path.
	 * @see Expression->alternative
```

File: src/SOFe/InfoAPI/Ast/InfoSegment.php [hunk 4]
```diff
@@ -31,30 +31,45 @@
...
+	 */
 	public Expression $head;
 
+	/**
+	 * @param string $string Content in a brace-pair among segments in InfoAPI template string.
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/InfoSegment.php:41]:
```suggestion
	 * @param string $string The substring inside braces
```

File: src/SOFe/InfoAPI/Ast/InfoSegment.php [hunk 5]
```diff
@@ -31,30 +31,45 @@
...
+	 * @param int $index Where does a path starts in its superstring.
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/InfoSegment.php:42]:
```suggestion
	 * @param int $index Starting index of `$string` in the template.
```

File: src/SOFe/InfoAPI/Ast/InfoSegment.php [hunk 6]
```diff
@@ -31,30 +31,45 @@
...
 			$index += strlen($pathString) + 1;
 		}
 		$self = new self;
 		$self->head = $head;
 		return $self;
 	}
 
+	/**
+	 * @param string $path Reminder: paths in a superstring are separated by "|".
+	 * @param int $index Where does a path starts in its superstring.
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/InfoSegment.php:67]:
```suggestion
	 * @param int $index Starting index of `$path` in the template string.
```

File: src/SOFe/InfoAPI/Ast/InfoSegment.php [hunk 7]
```diff
@@ -31,30 +31,45 @@
...
+	 */
 	public Expression $head;
 
+	/**
+	 * @param string $string Content in a brace-pair among segments in InfoAPI template string.
+	 * @param int $index Where does a path starts in its superstring.
+	 */
 	static public function parse(string $string, int $index) : self {
 		$head = new Expression;
 		$expr = $head;
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/InfoSegment.php:46]:
```suggestion
		// `$expr` is the tail of the linked list of expressions.
		$expr = $head;
```

File: src/SOFe/InfoAPI/Ast/InfoSegment.php [hunk 8]
```diff
@@ -31,30 +31,45 @@
...
 		$first = true;
 		foreach(explode("|", $string) as $pathString) {
 			if(!$first) {
 				$expr->alternative = new Expression;
+				// Support two or more coalescence paths:
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/InfoSegment.php:51]:
```suggestion
				// Update the tail of the linked list
```

File: src/SOFe/InfoAPI/Ast/InfoSegment.php [hunk 9]
```diff
@@ -31,30 +31,45 @@
...
 				$expr = $expr->alternative;
 			}
 			$first = false;
 			$expr->path = self::parsePath($pathString, $index);
 
+			// +1 to include the "|" of path separation:
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/InfoSegment.php:57]:
```suggestion
			// +1 to include the "|" separator
```

File: src/SOFe/InfoAPI/Ast/InfoSegment.php [hunk 10]
```diff
@@ -31,30 +31,45 @@
...
 			$index += strlen($pathString) + 1;
 		}
 		$self = new self;
 		$self->head = $head;
 		return $self;
 	}
 
+	/**
+	 * @param string $path Reminder: paths in a superstring are separated by "|".
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/InfoSegment.php:66]:
```suggestion
	 * @param string $path A path string separated by `|` in an info segment.
```

File: src/SOFe/InfoAPI/Ast/Segment.php [hunk 11]
```diff
@@ -23,6 +23,6 @@
 namespace SOFe\InfoAPI\Ast;
 
 /**
- * Either a `TextSegment` or an `InfoSegment`.
+ * Either a `TextSegment` or an `InfoSegment`. (Just like enums.)
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/Segment.php:26]: Is this comment really helpful?

File: src/SOFe/InfoAPI/Ast/TextSegment.php [hunk 12]
```diff
@@ -24,6 +24,7 @@
 
 /**
  * A segment of (escapes-resolved) plain-text.
+ * (The segment that InfoAPI ignore.)
```
Comment (SOF3) [inline src/SOFe/InfoAPI/Ast/TextSegment.php:27]: InfoAPI doesn't ignore it; InfoAPI still has to include this text.
```suggestion
 * A segment of (escapes-resolved) plain-text that gets appended to the output directly.
```

**Review by SOF3 — CHANGES_REQUESTED** [review]

File: src/SOFe/InfoAPI/Ast/ChildName.php (hunk 1 above)
Comment (Endermanbugzjfc) [reply src/SOFe/InfoAPI/Ast/ChildName.php:81]: I think the code states the former too. Why do you think that?

File: src/SOFe/InfoAPI/Ast/InfoSegment.php (hunk 7 above)
Comment (Endermanbugzjfc) [reply src/SOFe/InfoAPI/Ast/InfoSegment.php:46]: Thank you, was thinking if I should add this too.

File: src/SOFe/InfoAPI/Ast/Segment.php (hunk 11 above)
Comment (Endermanbugzjfc) [reply src/SOFe/InfoAPI/Ast/Segment.php:26]: It could act as a double-confirmation for people who know enum.

Comment (SOF3) [reply src/SOFe/InfoAPI/Ast/Segment.php:26]: I think this is just more confusing, especially when "enum" means different things in different languages. For example PHP enums are just constants.

Comment (Endermanbugzjfc) [reply src/SOFe/InfoAPI/Ast/Segment.php:26]:
```suggestion
 * Either a `TextSegment` or an `InfoSegment`.
```
👍 ?

**Comment by SOF3** [comment]
Closed since I rewrote the entire thing #29. Feel free to add more comments if the new code is confusing.

**Pull request closed without merging**